src/Processors/Formats/Impl/NativeORCBlockInputFormat.cpp

```diff
@@ -0,0 +1,1098 @@
+
+    size_t curr_offset = 0;
+    for (size_t i = 0; i < orc_str_column->numElements; ++i)
+    {
+        const auto * buf = orc_str_column->data[i];
+        if (buf)
+        {
+            size_t buf_size = orc_str_column->length[i];
+            column_chars_t.insert_assume_reserved(buf, buf + buf_size);
+            curr_offset += buf_size;
+        }
+
+        column_chars_t.push_back(0);
+        ++curr_offset;
+
+        column_offsets.push_back(curr_offset);
+    }
+    return {std::move(internal_column), std::move(internal_type), column_name};
+}
+
+static ColumnWithTypeAndName
+readColumnWithFixedStringData(const orc::ColumnVectorBatch * orc_column, const orc::Type * orc_type, const String & column_name)
+{
+    size_t fixed_len = orc_type->getMaximumLength();
+    auto internal_type = std::make_shared<DataTypeFixedString>(fixed_len);
+    auto internal_column = internal_type->createColumn();
+    PaddedPODArray<UInt8> & column_chars_t = assert_cast<ColumnFixedString &>(*internal_column).getChars();
+    column_chars_t.reserve(orc_column->numElements * fixed_len);
+
+    const auto * orc_str_column = dynamic_cast<const orc::StringVectorBatch *>(orc_column);
+    for (size_t i = 0; i < orc_str_column->numElements; ++i)
+    {
+        if (orc_str_column->data[i])
+            column_chars_t.insert_assume_reserved(orc_str_column->data[i], orc_str_column->data[i] + orc_str_column->length[i]);
+        else
+            column_chars_t.resize(column_chars_t.size() + fixed_len);
+    }
+
+    return {std::move(internal_column), std::move(internal_type), column_name};
+}
+
+
+template <typename DecimalType, typename BatchType, typename VectorType = ColumnDecimal<DecimalType>>
+static ColumnWithTypeAndName readColumnWithDecimalDataCast(
+    const orc::ColumnVectorBatch * orc_column, const orc::Type *, const String & column_name, DataTypePtr internal_type)
+{
+    using NativeType = typename DecimalType::NativeType;
+    static_assert(std::is_same_v<BatchType, orc::Decimal128VectorBatch> || std::is_same_v<BatchType, orc::Decimal64VectorBatch>);
+
+    auto internal_column = internal_type->createColumn();
+    auto & column_data = static_cast<VectorType &>(*internal_column).getData();
+    column_data.reserve(orc_column->numElements);
+
+    const auto * orc_decimal_column = dynamic_cast<const BatchType *>(orc_column);
+    for (size_t i = 0; i < orc_decimal_column->numElements; ++i)
+    {
+        DecimalType decimal_value;
+        if constexpr (std::is_same_v<BatchType, orc::Decimal128VectorBatch>)
+        {
+            Int128 int128_value;
+            int128_value.items[0] = orc_decimal_column->values[i].getLowBits();
+            int128_value.items[1] = orc_decimal_column->values[i].getHighBits();
+            decimal_value.value = static_cast<NativeType>(int128_value);
+        }
+        else
+            decimal_value.value = static_cast<NativeType>(orc_decimal_column->values[i]);
```

Comment:
```
/** Whether we can use memcpy instead of a loop with assignment to T from U.
  * It is Ok if types are the same. And if types are integral and of the same size,
  *  example: char, signed char, unsigned char.
  * It's not Ok for int and float.
  * Don't forget to apply std::decay when using this constexpr.
  */
template <typename T, typename U>
constexpr bool memcpy_can_be_used_for_assignment = std::is_same_v<T, U>
    || (std::is_integral_v<T> && std::is_integral_v<U> && sizeof(T) == sizeof(U));
```

Memcpy is not allowed when T is Decimal64 and U is int64_t